sdks/java/core/build.gradle

```diff
@@ -51,6 +51,8 @@ test {
 }
 
 dependencies {
+  // Required to loads constants from the model, e.g. max timestamp for global window
+  shadow project(path: ":beam-model-pipeline", configuration: "shadow")
```

Comment:
does it make sense for java/core to depend on model/pipeline?

I'd have expected a dependency arrow in the other direction based on their names and the little I know about what each does.

Reply:
Thanks for pointing this out. Yes, it is necessary to depend on beam-model-pipeline because it contains the proto pipeline definition. If we want to load the max timestamp from the proto we have to depend on it.

I agree, it's not very elegant and causes some issues which we still have to resolve (looking at the Java build failures).

Reply:
is the dependency being added here just to pick up the constant you've added?

or did java-core already have this dependency, but was (incorrectly) not making that explicit in its build.gradle (and picking it up transitively some other way)?

If it's the former, I am j/w if putting the constant in a proto designed for more "core" things would be better (answer could still be "that's not worth the overhead even though the dep graph would be a bit more logical")

Reply:
>is the dependency being added here just to pick up the constant you've added?

Just for picking up the constant at runtime.

>If it's the former, I am j/w if putting the constant in a proto designed for more "core" things would

What exactly do you mean? Couldn't be much more core than Runner API?

Reply:
I'm just trying to understand whether this is a case where:
- we're adding something (`X`) to a module `B`
- also adding a dep from `A` to `B` to let `A` use `X`
- but ideally `A` shouldn't depend on `B`, based on what `A` and `B` normally represent

an alternative might be to add `X` to a new module, and have `A` depend on that.

anyway, I'm missing a clear picture of what modules exist and should depend on what, feel free to disregard

Reply:
The goal is to have the max global window timestamp _once_ defined globally across all sdks. `beam-model-pipeline` is the proto definition which all SDKs depend on. 

In Python the module structure is much simpler and we just have the proto always available. In java, due to the runners being in java, we have a separate `beam-sds-java-core` module which so far doesn't depend on the pipeline module.

We have two options 1) depend on the pipeline definition in java-core 2) move GlobalWindow from java-core (I found this to be pretty tricky).

You propose to factor out the timestamp into a new module and let all sdks depend on that?

Reply:
thanks for that context!

"propose" is too strong; maybe if creating/maintaining more modules were free, then the leanest possible dependency structure would be nice, but I don't think that makes sense here. 

I mostly just wanted to talk through the module/dependency structure a bit. I probably need to dump it into a dot graph to really get it 😄 (aside: the existing visteg graphs i see generated [are fairly uninformative](https://cl.ly/153a4d17ce84/Screen%20Shot%202018-09-20%20at%2010.25.12%20AM.png))

anyway, no objections here

Reply:
backwards compatibility prevents us typically from changing package structure from what existed before and to move "constants" around.

I'm happy with this change especially since we know have vendored the proto/gRPC dependencies so this is unlikely to cause a problem for Java users. Not so certain how this impacts Python dependency wise.

Reply:
@ryan-williams Intersting, didn't know about the build dependency graphs.

@lukecwik Great. Do you have an idea what causes the `InvalidClassException` in the examples module? It must be due to the pipeline dependency.